dev/src/watch.ts

```diff
@@ -765,70 +431,435 @@ abstract class Watch {
+      .then(async () => {
+        if (!this.isActive) {
+          logger(
+            'Watch.onSnapshot',
+            this.requestTag,
+            'Not initializing inactive stream'
+          );
+          return;
+        }
+
+        await this.firestore.initializeIfNeeded();
+
+        const request: api.IListenRequest = {};
+        request.database = this.firestore.formattedName;
+        request.addTarget = this.getTarget(this.resumeToken);
+
+        // Note that we need to call the internal _listen API to pass additional
+        // header values in readWriteStream.
+        return this.firestore
+          .readWriteStream('listen', request, this.requestTag, true)
+          .then(backendStream => {
+            if (!this.isActive) {
+              logger(
+                'Watch.onSnapshot',
+                this.requestTag,
+                'Closing inactive stream'
+              );
+              backendStream.end();
+              return;
+            }
+            logger('Watch.onSnapshot', this.requestTag, 'Opened new stream');
+            this.currentStream = backendStream;
+            assert(this.currentStream !== undefined, 'CHECCK 1');
+            this.currentStream!.on('error', err => {
+              this.maybeReopenStream(err);
+            });
+            this.currentStream!.on('end', () => {
+              const err = new GrpcError('Stream ended unexpectedly');
+              err.code = GRPC_STATUS_CODE.UNKNOWN;
+              this.maybeReopenStream(err);
+            });
+            this.currentStream!.pipe(this.stream);
+            this.currentStream!.resume();
+          });
+      })
+      .catch(err => {
+        this.closeStream(err);
+      });
+  }
+
+  /**
+   * Checks if the current target id is included in the list of target ids.
+   * If no targetIds are provided, returns true.
+   */
+  private affectsTarget(
+    targetIds: number[] | undefined,
+    currentId: number
+  ): boolean {
+    if (targetIds === undefined || targetIds.length === 0) {
+      return true;
+    }
+
+    for (const targetId of targetIds) {
+      if (targetId === currentId) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * Assembles a new snapshot from the current set of changes and invokes the
+   * user's callback. Clears the current changes on completion.
+   */
+  private pushSnapshot(
+    readTime: Timestamp,
+    nextResumeToken?: Uint8Array
+  ): void {
+    console.warn('push snapshot readtime: ', readTime);
+    const changes = this.extractChanges(readTime);
+    const appliedChanges = this.computeSnapshot(readTime);
+
+    if (!this.hasPushed || appliedChanges.length > 0) {
+      logger(
+        'Watch.onSnapshot',
+        this.requestTag,
+        'Sending snapshot with %d changes and %d documents',
+        String(appliedChanges.length),
+        this.docTree.length
+      );
+      // Pass doctree in as a const, so we don't lose reference to 'this'.
```

Comment:
That's not really the reason we are doing this. We want to always refer to what `this.docTree` points to, and not to what the `docTree` in `this` points to. Maybe make this something like:

```
// We pass the current set of changes, even if `docTree` is modified later.
const currentState = this.docTree;
```